src/gc/proxy.d

```diff
@@ -68,132 +103,159 @@ extern (C)
         // NOTE: Due to popular demand, this has been re-enabled.  It still has
         //       the problems mentioned above though, so I guess we'll see.
 
-        instance.collectNoStack(); // not really a 'collect all' -- still scans
-                                    // static data area, roots, and ranges.
+        if (instance !is null)
+        {
+            instance.collectNoStack(); // not really a 'collect all' -- still scans
+                                        // static data area, roots, and ranges.
 
-        thread_term();
+            thread_term();
 
-        ManualGC.finalize(instance);
-        ConservativeGC.finalize(instance);
+            ManualGC.finalize(instance);
+            ConservativeGC.finalize(instance);
+        }
     }
 
     void gc_enable()
     {
+        gc_init_nothrow();
         instance.enable();
     }
 
     void gc_disable()
     {
+        assert(instance !is null);
         instance.disable();
```

Comment:
Will this break code that calls `GC.disable` at the beginning of the program?  Sometimes it's useful to do this and schedule your own calls to `GC.collect`; I've done this before to optimize GC performance.

Would it be better to also initialize the GC in `gc_enable` / `gc_disable`, in addition to doing it upon the first memory allocation? IOW, initialize the GC the first time it's used in some way, whether it's a memory allocation or enable/disable, etc..

Reply:
If there is a use case for it, then yes - it should be initialized.
Which of these methods do you think should initialize the GC?

Reply:
Definitely GC.disable as it's fairly common. Don't think anyone has explicitly called `GC.enable` yet. Even better if we could memorize GC.disable and lazily initialize and disable the GC if it's used later on. It would be weird to initialize a GC just to disable it.

Reply:
Hm... just had a weird idea.

What if the instance is initialized with a GC that when first used, figures out what GC to allocate, sets the 'instance' to the new one, and calls the appropriate function on it. Then we can handle things like a bool to track the disabling without having to initialize until someone tries to call malloc, etc.

Then once the *real* instance is in place, there is no longer any checking for initialization.

You'd still need to atomically load the instance every time you use it, unfortunately.

Reply:
I like this idea.

Reply:
>Don't make things too complicated though, when in doubt just initialize the GC on any access.

let's do that now, refine later

Reply:
@schveiguy it's nice - the almost-null-object pattern? :) - but a bit too cute. I'm not worried about optimizing the call to disable on an uninitialized gc. Let's move forward with the dclp, if you concurrently arrive at a better solution we can review that. Thx!

Reply:
@andralex dclp is still required, actually. When it replaces itself, it should only do so once. In essence, it's not altering the requirements that this PR is solving. What it would do, however, is avoid any the checks for null on every call once the real GC is in place. Not much cost, but still something. I'm just thinking about what @rainers said about the performance.

I think it also would make the code look cleaner. I'm interested to see if @marler8997 can pull it off, and how it does look.

But in any case, this PR is good enough IMO. Just a thought I had.

Reply:
> Just a thought I had.

Possible @rainers had the same thought, I just didn't see it 😄 

> An alternative would be to implement the GCInterface with an object that switches the instance global to the standard GC if necessary.

Reply:
What are the methods requiring that behavior `disable`, `addRoot, removeRoot, rootIter`, `addRange, removeRange, rangeIter`? Sounds like a reasonable approach for lazy initialization.

Reply:
BTW, that just requires an `atomicLoad!(MemoryOrder.acq)`, and that's free on strongly-ordered CPUs.

Reply:
I would initialize a full GC on anything that requires storing an array of things to do, because that involves allocation. Just having a global bool is OK to cache until you need the whole GC.

> that's free on strongly-ordered CPUs

Right, x86 doesn't have issues with loads of a word, so those will be fine. It's the other architectures I don't know enough about how the performance might be. I would leave that decision/discussion to those who know more :)

Reply:
> I would initialize a full GC on anything that requires storing an array of things to do

Hm... I guess if you need to allocate using the C heap you can, as long as the behavior matches what the GC would actually do. You might even pass those into the GC initializer, and it would just take over ownership of the arrays.

Reply:
>BTW, that just requires an atomicLoad!(MemoryOrder.acq), and that's free on strongly-ordered CPUs.

Please KISS (Keep It Simple and ConServative) for now. Then we'll take a separate PR with loads and stores tightened. If we ever have trouble with those we'll be able to track that easily. Thanks!